### PR TITLE
HDDS-5151. Support ByteBuffer read in OzoneInputStream

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-5151

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 1]
```diff
@@ -507,4 +509,73 @@ private void handleReadError(IOException cause) throws IOException {
...
+      }
+      totalReadLen += numBytesRead;
+      bufferLen -= numBytesRead;
+      if (current.getRemaining() <= 0 &&
+          ((chunkIndex + 1) < chunkStreams.size())) {
+        chunkIndex += 1;
+      }
+    }
+    return totalReadLen;
+  }
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:580]: Can we avoid duplicating logic from `read(byte[], ...)`?  (Also in `ChunkInputStream` and `KeyInputStream`.)

**Review by adoroszlai — COMMENTED** [review]
Thanks @ChenSammi for this improvement.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java]: Good point.  I have updated a new patch for this.

**Comment by jojochuang** [comment]
@hanishakoneru or @bshashikant should probably check this out too. Changes the client side inputstream implementation

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java]: Thanks @ChenSammi for the update.  The new implementation using strategy pattern looks good.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/read/TestKeyInputStream.java [hunk 2]
```diff
@@ -351,4 +362,25 @@ private void assertReadFully(byte[] data, InputStream in,
     }
     Assert.assertEquals(data.length, totalRead);
   }
+
+  private void assertReadFullyUseByteBuffer(byte[] data, KeyInputStream in,
```
Comment (hanishakoneru) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/read/TestKeyInputStream.java:366]: NIT: assertReadFullyUsingByteBuffer

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/read/TestKeyInputStream.java [hunk 3]
```diff
@@ -200,6 +201,8 @@ public void testReadChunk() throws Exception {
       for (int bufferSize : bufferSizeList) {
         assertReadFully(data, keyInputStream, bufferSize, 0);
         keyInputStream.seek(0);
+        assertReadFullyUseByteBuffer(data, keyInputStream, bufferSize, 0);
+        keyInputStream.seek(0);
```
Comment (hanishakoneru) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/read/TestKeyInputStream.java:205]: Can we create a new test for ByteBuffer read?

**Review by hanishakoneru — COMMENTED** [review]
Thank you @ChenSammi for working on this. Patch LGTM overall. I have some minor comments posted inline.

**Comment by ChenSammi** [comment]
@hanishakoneru thanks the detail code review,  I have addressed all the comments in the new commit.

**Comment by ChenSammi** [comment]
@hanishakoneru and @adoroszlai ,  would you help to take another look again?

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ReaderStrategy.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+package org.apache.hadoop.hdds.scm.storage;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * A Reader interface to work with InputStream.
+ */
+public interface ReaderStrategy {
```
Comment (hanishakoneru) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ReaderStrategy.java:27]: Similar to implementations of this interface, do you think we can rename to something like ByteReader?

**Review by hanishakoneru — COMMENTED** [review]
Thanks @ChenSammi for addressing the comments. Sorry I missed a few last time.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ReaderStrategy.java (hunk 4 above)
Comment (ChenSammi) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ReaderStrategy.java]: Sure.

**Comment by hanishakoneru** [comment]
Thank you @ChenSammi for working on this. 
LGTM. +1 for merge.

**Review by hanishakoneru — APPROVED** [review]

**Comment by ChenSammi** [comment]
Thanks @hanishakoneru @adoroszlai @jojochuang for the code review.

**Pull request merged**